File: 01_Arrays_&_Hashing/09_Longest_Consecutive_Sequence/0128-longest-consecutive-sequence.py
"""
Problem: LeetCode 128 -Longest Consecutive Sequence

Key Idea:
To find the longest consecutive subsequence, we first create a set of all the elements in the input array 'nums'. Then, for each element in the array, we check if it is the starting element of a consecutive subsequence. To do this, we check if the element before the current element (i.e., nums[i] - 1) exists in the set. If it doesn't, it means nums[i] is the starting element of a consecutive subsequence. From here, we keep incrementing the current element until the consecutive subsequence ends (i.e., the next element does not exist in the set). We keep track of the length of the consecutive subsequence and update the maximum length found so far.

Time Complexity:
The time complexity of this solution is O(n), where n is the number of elements in the input array 'nums'. The set operations (addition and lookup) take O(1) time on average, and we perform these operations for each element once.

Space Complexity:
The space complexity is O(n), where n is the number of elements in the input array 'nums'. In the worst case, the set will store all elements from the input array if they are all distinct.
"""


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        num_set = set(nums)
        max_length = 0

        for num in num_set:
            if num - 1 not in num_set:
                current_num = num
                current_length = 1

                while current_num + 1 in num_set:
                    current_num += 1
                    current_length += 1

                max_length = max(max_length, current_length)

        return max_length

File: 01_Arrays_&_Hashing/09_Longest_Consecutive_Sequence/0128-longest-consecutive-sequence.rs
/*
Problem: LeetCode 128 - Longest Consecutive Sequence

Key Idea:
The key idea is to use a hash set to efficiently keep track of the consecutive sequence lengths starting from each number.

Approach:
1. Create a hash set to store the numbers from the input array.
2. Initialize a variable `longest_streak` to track the maximum consecutive sequence length found so far.
3. Iterate through the numbers in the input array:
   - For each number `num`, check if `num - 1` exists in the hash set. If not, it's a potential starting point of a consecutive sequence.
   - While `num + 1` exists in the hash set, increment `num` to extend the sequence.
   - Update `longest_streak` with the maximum of its current value and the length of the current sequence.
4. Return `longest_streak` as the result.

Time Complexity:
O(n), where n is the number of elements in the input array. We iterate through the array once and perform O(1) operations on average for each number.

Space Complexity:
O(n), as the hash set can contain all the elements from the input array.
*/

use std::collections::HashSet;

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let num_set: HashSet<i32> = nums.iter().cloned().collect();
        let mut longest_streak: i32 = 0;

        for num in num_set.iter() {
            if !num_set.contains(&(num - 1)) {
                let mut streak = 1;
                let mut current = *num;

                while (num_set.contains(&(current + 1))) {
                    current += 1;
                    streak += 1;
                }

                longest_streak = std::cmp::max(streak, longest_streak);
            }
        }

        longest_streak
    }
}
